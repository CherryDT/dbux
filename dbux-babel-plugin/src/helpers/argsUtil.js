import EmptyArray from '@dbux/common/src/util/EmptyArray';


function makeSpreadableArgumentCfg(argPath) {
  return {
    isSpread: argPath.isSpreadElement()
  };
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `CallExpression`, `ArrayExpression`.
 */
export function makeSpreadableArgumentArrayCfg(argumentPaths) {
  return argumentPaths?.map(makeSpreadableArgumentCfg) || EmptyArray;
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `ObjectExpression`.
 */
export function makeSpreadableArgumentObjectCfg(propertyPaths) {
  return propertyPaths?.map((propPath) => ({ 
    key: propPath.node.key,
    isSpread: propPath.isSpreadElement()
  })) || EmptyArray;
}
