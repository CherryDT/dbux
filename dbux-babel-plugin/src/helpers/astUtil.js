// import * as t from '@babel/types';

/**
 * If true, given node was injected/instrumented by babel or some plugin,
 * meaning it does not exist in original source code and (usually) does not deserve further analysis.
 */
export function isNodeInstrumented(node) {
  return !node.loc;
}

export function isPathInstrumented(path) {
  return !path.node?.loc;
}

export function getPathContextId(path) {
  return path.getData('staticId');
}

const scopeBreakingNodeTypes = Object.fromEntries([
  'ReturnStatement',
  'ThrowStatement',
  'BreakStatement',
  'ContinueStatement'
].map(type => ([type, true])));

/**
 * @return Whether execution cannot continue right after the given instruction.
 */
export function doesNodeEndScope(node) {
  return !!scopeBreakingNodeTypes[node.type];
}

export function isPrivateClassMember(node) {
  switch (node.type) {
    case "ClassPrivateProperty":
    case "ClassPrivateMethod":
      return true;
  }
  return false;
}