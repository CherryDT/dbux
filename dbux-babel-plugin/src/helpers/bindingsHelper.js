/**
 * Bindings usually refer to "creations" of variables, 
 * such as declarations, parameters, module exports etc.
 * 
 * @file
 * 
 * class `Binding`:
 * @see https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/binding.ts#L5
 * 
 * `registerDeclaration`:
 * @see https://github.com/babel/babel/blob/672a58660f0b15691c44582f1f3fdcdac0fa0d2f/packages/babel-traverse/src/scope/index.ts#L646
 * 
 * binding collision logic:
 * @see https://github.com/babel/babel/blob/672a58660f0b15691c44582f1f3fdcdac0fa0d2f/packages/babel-traverse/src/scope/index.ts#L515
 */

// import { logInternalError } from '../log/logger';
// import EmptyObject from '@dbux/common/src/util/EmptyObject';
// import { isInLoc1D } from './locHelpers';
// import { isNodeInstrumented } from './astUtil';

// /**
//  * Find only bindings which have at least one non-instrumented path;
//  * i.e. variables that exist in the original source code.
//  */
// export function* iterateBindings(path) {
//   if (Array.isArray(path)) {
//     for (const pathEntry of path) {
//       yield* iterateBindings(pathEntry);
//     }
//     return;
//   }

//   const bindings = path.scope?.bindings;
//   if (!bindings) {
//     logInternalError(`[iterateRealVariableBindings] argument is not path or has no bindings: ${path.toString()}`);
//   }
//   else {
//     yield* Object.keys(bindings)
//       .filter(varName => 
//         !isNodeInstrumented(bindings[varName].identifier) ||
//         bindings[varName].referencePaths.some(varPath => !isNodeInstrumented(varPath.node)))
//       .map(varName => ([varName, bindings[varName]]));
//   }
// }

// // export function getAllBoundVariable(path) {
// //   return Array.from(iterateBindings(path)).map(([name]) => name);
// // }

// export function getClosestScopedPath(path) {
//   let current = path;
//   do {
//     if (current.scope) {
//       return current;
//     }
//     current = path.parentPath;
//   }
//   while (current);

//   return null;
// }

// /**
//  * 
//  */
// export function iterateVarAccessInLoc1D(path, containingLoc1D, cb) {
//   const scopedPath = getClosestScopedPath(path);
//   const bindings = scopedPath?.scope?.bindings || EmptyObject;
//   const varNames = Object.keys(bindings);

//   // see: https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#bindings
//   for (const varName of varNames) {
//     const binding = bindings[varName];
//     const {
//       referencePaths,
//       identifier,
//       constantViolations
//     } = binding;

//     if (isNodeInstrumented(identifier)) {
//       // skip: variable injected by instrumentation, not in original source code
//       continue;
//     }

//     if (isInLoc1D(identifier, containingLoc1D)) {
//       // add declaration
//       cb(identifier.name, identifier, false);
//     }

//     // add all accesses
//     // WARNING: `referencePaths` only contains reads of given indentifier in given scope. It does not seem like they can ever count as constantViolations.
//     for (const varPath of referencePaths) {
//       if (isInLoc1D(varPath.node, containingLoc1D)) {
//         // add var access

//         // TODO: const initializer is "write" as well (but not a violation)

//         const isWrite = !constantViolations.includes(varPath);
//         // yield { varPath, name: identifier.name, isWrite };
//         cb(identifier.name, varPath, isWrite);
//       }
//     }
//   }
// }

// export function getRealVariableNamesInLoc1D(path, loc1D) {
//   const names = [];
//   iterateVarAccessInLoc1D(path, loc1D, (name) => names.push(name));
//   return names;
// }

/**
 * @return {Binding} The binding that creates/declares the given identifier (if it can be found).
 */
export function getBinding(idPath) {
  crawl(idPath);
  // NOTE: there is also `getAllBindings`
  return idPath.scope.getBinding((idPath.node).name);
}

/**
 * Crawl: "Manually reprocess this scope to ensure that the moved params are updated."
 * @see https://github.com/babel/babel/blob/672a58660f0b15691c44582f1f3fdcdac0fa0d2f/packages/babel-traverse/src/scope/index.ts#L863
 */
function crawl(path) {
  // NOTE: only need to crawl after bindings have changed, so probably not necessary
  // path.scope.crawl();
}

/**
 * 
 * @returns {Path} The path that creates/declares the given identifier.
 */
export function getBindingPath(idPath) {
  return getBinding(idPath)?.path;
}