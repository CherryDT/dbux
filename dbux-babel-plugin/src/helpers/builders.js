/**
 * Helpers to generate partial AST's.
 */

import { parse } from '@babel/parser';
import { codeFrameColumns } from "@babel/code-frame";
import traverse from "@babel/traverse";
import * as t from "@babel/types";

export function buildTryFinally(tryNodes, finallyNodes) {
  return t.tryStatement(t.blockStatement(tryNodes), null, t.blockStatement(finallyNodes))
}

/**
 * Build AST from source through @babel/parser
 * @param {*} source 
 */
export function buildSource(source) {
  let ast;
  try {
    source = `${source}`;
    ast = parse(source);
  } catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message +=
        "\n" +
        codeFrameColumns(source, {
          start: {
            line: loc.line,
            column: loc.column + 1,
          },
        });
    }
    throw err;
  }

  const nodes = ast.program.body;
  nodes.forEach(n => traverse.removeProperties(n));
  return nodes;
}

// https://stackoverflow.com/questions/35925798/how-to-add-an-import-to-the-file-with-babel