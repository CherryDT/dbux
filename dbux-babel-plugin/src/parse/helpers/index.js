import Declaration from './Declaration';

export {
  Declaration
};