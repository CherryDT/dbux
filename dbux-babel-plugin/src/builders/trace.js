import template from '@babel/template';
import * as t from '@babel/types';
import { newLogger } from '@dbux/common/src/log/logger';
import { getPresentableString } from '../helpers/pathHelpers';
import { bindTemplate } from '../helpers/templateUtil';

// eslint-disable-next-line no-unused-vars
const { log, debug, warn, error: logError } = newLogger('builders/trace');

const Verbose = 2;

export const buildTraceId = bindTemplate(
  '%%traceId%% = %%makeTraceId%%(%%staticTraceId%%)',

  function buildTraceId(templ, state, traceIdVar, inProgramStaticTraceId) {
    // TODO: add custom trace data
    const { ids: { aliases: {
      makeTraceId
    } } } = state;

    return templ({
      makeTraceId,
      staticTraceId: t.numericLiteral(inProgramStaticTraceId),
      traceId: traceIdVar
    });
  }
);

export const buildTraceExpression = bindTemplate(
  '%%traceExpression%%(%%expr%%, %%tid%%)',
  function buildTraceExpression(templ, path, state, traceType) {
    // TODO: add custom trace data
    // const { scope } = path;
    const { ids: { aliases: {
      traceExpression
    } } } = state;

    const tid = buildTraceId(path, state, traceType);
    const expr = path.node;
    Verbose && debug('[te]', getPresentableString(path));

    // NOTE: templates only work on `Node`, not on `NodePath`, thus they lose all path-related information.
    
    // TODO: keep `path` data etc, if necessary - `onCopy(path, newPath);`

    return templ({
      traceExpression,
      expr,
      tid
    });
  }
);

export const buildTraceNoValue = function buildTraceNoValue(templ, path, state, traceType) {
  const { ids: { dbux } } = state;
  const traceId = state.traces.addTrace(path, traceType);
  // console.warn(`traces`, state.traces);
  return templ({
    dbux,
    traceId: t.numericLiteral(traceId)
  });
}.bind(null, template('%%dbux%%.t(%%traceId%%)'));

