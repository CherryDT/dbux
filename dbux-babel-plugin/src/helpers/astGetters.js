/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';
import {isDebug} from 'dbux-common/src/util/nodeUtil';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}

/**
 * 
 */
export function getAllClassParents(path) {
  const classParents = [];
  let classParent = path;
  do {
    classParent = classParent.findParent(p => t.isClass(p.node));
    if (classParent) {
      classParents.push(classParent);
    }
  } while (classParent);
  return classParents;
}