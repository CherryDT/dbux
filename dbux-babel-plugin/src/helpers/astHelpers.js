/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}