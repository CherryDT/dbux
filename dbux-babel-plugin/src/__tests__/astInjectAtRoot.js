import { runSnapshotTests } from '../testing/test-util';
import { buildSource } from '../helpers/builders';

const codes = [
  `
function f2() {
  AddCustomCode();
  AddCustomCode();
}
f2();
f2();
`,
  `
function f1() {
  console.log('hi');
}

f1();
`
];

const plugin = function ({ types: t }) {
  const customCode = buildSource(`someFancyCodeHere();`);

  let imported = false;
  let root;
  return {
    visitor: {
      Program(path) {
        root = path;
      },
      CallExpression(path) {
        if (!imported && path.node.callee.name === "AddCustomCode") {
          // add import if it's not there
          imported = true;
          root.unshiftContainer('body', customCode);
          root.resync();
        }
      }
    }
  };
};

codes.forEach((code, i) => {
  const title = `[${i}]${__filename}`;
  runSnapshotTests(code, __filename, title, {
    plugin
  });
});
