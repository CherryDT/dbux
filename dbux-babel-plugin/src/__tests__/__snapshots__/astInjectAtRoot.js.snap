// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbux-babel-plugin astInjectAtRoot.es5.js: astInjectAtRoot.es5.js 1`] = `
"
/* #################################################################################### */

function f2() {
  AddCustomCode();
  AddCustomCode();
}
f2();
f2();

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

someFancyCodeHere();

/* #################################################################################### */
function f2() {
  AddCustomCode();
  AddCustomCode();
}

f2();
f2();
"
`;

exports[`dbux-babel-plugin astInjectAtRoot.es5.js: astInjectAtRoot.es5.js 2`] = `
"
/* #################################################################################### */

function f1() {
  console.log('hi');
}

f1();

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

/* #################################################################################### */
function f1() {
  console.log('hi');
}

f1();
"
`;

exports[`dbux-babel-plugin astInjectAtRoot.esNext.js: astInjectAtRoot.esNext.js 1`] = `
"
/* #################################################################################### */

function f2() {
  AddCustomCode();
  AddCustomCode();
}
f2();
f2();

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

/* #################################################################################### */
function f2() {
  AddCustomCode();
  AddCustomCode();
}

f2();
f2();
"
`;

exports[`dbux-babel-plugin astInjectAtRoot.esNext.js: astInjectAtRoot.esNext.js 2`] = `
"
/* #################################################################################### */

function f1() {
  console.log('hi');
}

f1();

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

/* #################################################################################### */
function f1() {
  console.log('hi');
}

f1();
"
`;
